batch2/batch/front_end/front_end.py

```diff
@@ -444,14 +446,22 @@
                 secrets = spec.get('secrets')
                 if not secrets:
                     secrets = []
-                    spec['secrets'] = secrets
+                spec['secrets'] = secrets
                 secrets.append({
                     'namespace': BATCH_PODS_NAMESPACE,
                     'name': userdata['gsa_key_secret_name'],
                     'mount_path': '/gsa-key',
                     'mount_in_copy': True
                 })
 
+                for secret in secrets:
+                    if secret['name'] != userdata['gsa_key_secret_name'] and user != 'ci':
+                        raise web.HTTPBadRequest(reason=f'unauthorized secret {secret["name"]}')
+
+                sa = spec.get('service_account')
+                if sa and sa['name'] != ksa_name and user != 'ci':
+                    raise web.HTTPBadRequest(reason=f'unauthorized service account name {sa["name"]}')
```

Comment:
I think CI runs jobs with a variety of service accounts right now.  For purposes of isolation, that's probably a good idea.  
I think we can let CI do roughly what it wants, and the main point is to lock down non-CI users.  So users should mount their gsa-key, tokens and authorize their service account (if they have one), but can't mount custom secrets or use other service accounts than their own.  CI can do what it wants.

We could also make all secrets and service accounts explicit so users don't risk exposing secrets to jobs that don't need them.  Pipeline and batch could have different defaults.

Reply:
There isn't a field in the jwt to know what secret the user tokens are stored in.

Reply:
Just being pedantic:
 - There are no jwt anymore.  There are tokens.  Tokens are a dict(str, str) from namespace to the session ID (an opaque random string) for that namespace.
 - hailtop.auth has {async_}get_user_info that calls the auth service with the token to get the user info from the database.
 - this contains a jwt_secret_name (now misnamed, I rename it in my recent user admin PR) that contains the name of the secret storing the tokens.
 - The authentication decorators also call auth with the incoming token to get the full user record which is passed to the endpoint handlers as userdata.  There is where we get the user's gsa-key: see /jobs/create.

Reply:
Ok. thanks. this makes more sense now.

The other problem I'm running into is the batch test `test_service_account` is trying to use the ci-agent service account, but the job is submitted by the test user. I'm looking into making a test service account and adding that service account name to the test user tokens (I believe I can do this with the code in the create_accounts build step).